module/zfs/zfs_vnops.c

```diff
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+		 */
+		if (BP_IS_PROTECTED(&bps[0])) {
+			if (srczfsvfs != dstzfsvfs) {
+				dmu_tx_abort(tx);
+				kmem_free(bps, sizeof (bps[0]) * nbps);
+				error = SET_ERROR(ENODEV);
+				break;
+			}
+		}
+
+		dmu_tx_hold_sa(tx, dstzp->z_sa_hdl, B_FALSE);
+		db = (dmu_buf_impl_t *)sa_get_db(dstzp->z_sa_hdl);
+		DB_DNODE_ENTER(db);
+		dmu_tx_hold_write_by_dnode(tx, DB_DNODE(db), dstoffset, size);
+		DB_DNODE_EXIT(db);
+		zfs_sa_upgrade_txholds(tx, dstzp);
+		error = dmu_tx_assign(tx, TXG_WAIT);
+		if (error != 0) {
+			dmu_tx_abort(tx);
+			kmem_free(bps, sizeof (bps[0]) * nbps);
+			break;
+		}
+
+		/*
+		 * Copy source znode's block size. This only happens on the
+		 * first iteration since zfs_rangelock_reduce() will shrink down
+		 * lr_length to the appropriate size.
+		 */
+		if (dstlr->lr_length == UINT64_MAX) {
+			zfs_grow_blocksize(dstzp, srcblksz, tx);
+			zfs_rangelock_reduce(dstlr, dstoffset, length);
+		}
+
+		dmu_brt_addref(dstzfsvfs->z_os, dstzp->z_id, dstoffset, size,
+		    tx, bps, nbps);
+
+		zfs_clear_setid_bits_if_necessary(dstzfsvfs, dstzp, cr,
+		    &clear_setid_bits_txg, tx);
+
+		zfs_tstamp_update_setup(dstzp, CONTENT_MODIFIED, mtime, ctime);
+
+		/*
+		 * Update the file size (zp_size) if it has changed;
+		 * account for possible concurrent updates.
+		 */
+		while ((dstsize = dstzp->z_size) < dstoffset + size) {
+			(void) atomic_cas_64(&dstzp->z_size, dstsize,
+			    dstoffset + size);
+		}
+
+		error = sa_bulk_update(dstzp->z_sa_hdl, bulk, count, tx);
+
+		zfs_log_clone(zilog, tx, TX_CLONE, dstzp, dstioflag, dstoffset,
+		    size, srcblksz, bps, nbps);
+		dmu_tx_commit(tx);
+
+		kmem_free(bps, sizeof (bps[0]) * nbps);
+
+		if (error != 0)
+			break;
+
+		srcoffset += size;
+		dstoffset += size;
+		length -= size;
+		*donep += size;
+	}
+
+	zfs_znode_update_vfs(dstzp);
+	zfs_rangelock_exit(dstlr);
+	zfs_rangelock_exit(srclr);
+
+	/*
+	 * If we made no progress, return error.
+	 * Otherwise, it's at least a partial write, so it's successful.
+	 */
+	if (*donep == 0) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (error);
+	}
+
+	if (dstioflag & (O_SYNC | O_DSYNC) ||
+	    dstzfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS) {
+		zil_commit(zilog, dstzp->z_id);
+	}
+
+	/*
+	 * TODO: We don't log access time update to ZIL of the source
+	 *       file system.
+	 */
+	ZFS_ACCESSTIME_STAMP(srczfsvfs, srczp);
+	zfs_exit_two(srczfsvfs, dstzfsvfs);
+
+	return (0);
+}
+
+int
+zfs_clone_range_replay(znode_t *zp, uint64_t offset, uint64_t length,
+    uint64_t blksz, const blkptr_t *bps, size_t nbps)
+{
+	zfsvfs_t	*zfsvfs;
+	dmu_buf_impl_t	*db;
+	dmu_tx_t	*tx;
+	int		error;
+	int		count = 0;
+	sa_bulk_attr_t	bulk[4];
+	uint64_t	mtime[2], ctime[2];
+
+	ASSERT3U(offset, <, MAXOFFSET_T);
+	ASSERT3U(length, >, 0);
+	ASSERT3U(nbps, >, 0);
+
+	zfsvfs = ZTOZSB(zp);
+
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(zp);
+
+	ASSERT(zfsvfs->z_replay);
+	ASSERT(!zfs_is_readonly(zfsvfs));
+
+	if (!spa_feature_is_enabled(dmu_objset_spa(zfsvfs->z_os),
```

Comment:
For the ZIL logging of xattr=sa (zfs_replay_setxsaattr)
we have an ASSERT(spa_feature_is_active()), since,
we should definitely not be logging TX_CLONE records
if we haven't activated the feature.

Reply:
Done.